### PR TITLE
WebSocket implementation of streaming API

### PR DESCRIPTION
**Pull request by 178inaba**
This PR provides a WebSocket implementation of the streaming API.

### CODE REVIEW

**Comment by mattn** [comment]
Thanks for your great PR. I don't understand mastodon websocket well. Do you think mastodon.Client should inherit websocket client? Not an another client like WSClient? I don't have objection for this PR.

**Comment by 178inaba** [comment]
WebSocket implementation also needed AccessToken, so I thought that it would be okay for the client to inherit the WebSocket client.
But I began to think that it might be better to separate clients as you say.
When dividing clients, I think that you need a function to return the following WebSocket client which takes over Server and AccessToken.

```go
type WSClient struct {
	websocket.Dialer
	Server      string
	AccessToken string
}

func (c *Client) NewWSClient() WSClient {
	return WSClient{
		Server:      c.config.Server,
		AccessToken: c.config.AccessToken,
	}
}
```

I am prepared to implement this.

Is it better to split it into Client and WSClient?

**Comment by mattn** [comment]
Thank you for your suggestion. I'm thinking accessor for WSClient should be private.

```go
type WSClient struct {
	websocket.Dialer
	client      Client
}

func (c *Client) NewWSClient() *WSClient {
	return &WSClient{
		client: c,
	}
}
```

**Comment by 178inaba** [comment]
Good suggestion! I will implement it accordingly.

**Comment by 178inaba** [comment]
Completed!

File: streaming_ws_test.go [hunk 1]
```diff
@@ -234,7 +234,7 @@ func TestHandleWS(t *testing.T) {
 }
 
 func TestDialRedirect(t *testing.T) {
-	client := NewClient(&Config{})
+	client := NewClient(&Config{}).NewWSClient()
 	_, err := client.dialRedirect(":")
 	if err == nil {
```
Comment (mattn) [inline streaming_ws_test.go]: Maybe, This is problem of my test code. When func return nil for err expected, it should check `if err != nil` and message should be `should not be fail`. When func return NOT nil for err expected, it should check `if err == nil` and message should be `should be fail`.

Comment (178inaba) [inline streaming_ws_test.go]: I understand.
However, this line seems not to be wrong.
It wrong is the [264-265](https://github.com/mattn/go-mastodon/pull/39/files/010ec2eaf92da466640e155dfac5a9611b6a0f33#diff-7f77780830ccced377be5bedb5727d5aR264) line?

File: streaming_ws_test.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+	client := NewClient(&Config{}).NewWSClient()
+	_, _, err := client.dial(":")
+	if err == nil {
+		t.Fatalf("should be fail: %v", err)
+	}
+
+	_, rawurl, err := client.dial("ws://" + ts.Listener.Addr().String())
+	if err == nil {
+		t.Fatalf("should not be fail: %v", err)
```
Comment (178inaba) [inline streaming_ws_test.go]: Is it correct to modify the message on this line to "`should be fail`"?

Comment (mattn) [inline streaming_ws_test.go]: YEP! :+1:

Comment (178inaba) [inline streaming_ws_test.go]: OK!
I'll do it right now!

**Comment by 178inaba** [comment]
Fixed!
Also, since the error message was wrong with another file, I corrected it and issued PR.
https://github.com/mattn/go-mastodon/pull/42

**Comment by mattn** [comment]
Thanks!

**Pull request merged**